setwd('~/Projects/03_MICOM/')

# scripts used
source('git/scripts/whitehead.accession.uniprot.map.R')
source('git/scripts/accession.uniprot.get.R')
source('git/scripts/map_uniprot_to_gene.R')
source('git/scripts/accession.uniprot.get.R')
source('git/scripts/description.keyword.R')
source('git/scripts/hgnc_aliases.R')
source('git/scripts/standardize_bait_names.R')
source('git/scripts/extract_ip_accession.R')

# write from
writefrom = 'run050'
writeto = 'run061'


# load non-refined basic dataset from subsetted raw proteomic data
# run006 contains the raw data, as well as logFC between releveant 
# intensity values for iTRAQ replicates.
paths = list.files(paste0('data/genoppi_input/',writefrom), full.names = T)  # old: paths = list.files('data/genoppi_input/run006', full.names = T) 
paths.whitehead = !grepl('Broad', paths)
paths.broad = grepl('Broad', paths)

# download hgnc mapping
hgnc <- read.delim(url("https://www.genenames.org/cgi-bin/download/custom?col=gd_hgnc_id&col=gd_app_sym&col=gd_app_name&col=gd_prev_sym&col=gd_aliases&col=gd_pub_acc_ids&col=gd_pub_refseq_ids&col=gd_name_aliases&col=gd_pub_ensembl_id&col=md_prot_id&col=gd_pub_eg_id&col=gd_prev_name&hgnc_dbtag=on&order_by=gd_app_sym_sort&format=text&submit=submit"))
#write.table(extract_ip_accession(list.files('~/Projects/03_MICOM/data/genoppi_input/run050/', full.names = T)), '~/Desktop/201119_micom_accession_numbers.tsv', quote = F, row.names = F)
uniprot_to_gene <- read.delim(url('https://www.uniprot.org/mapping/M202011205C475328CEF75220C360D524E9D456CE04FF656.tab')) # generated by extracting accession numbers from all files an quering with uniprot (see above line).
#write.csv(uniprot_to_gene, '29MAY20_micpm_uniprot_to_gene_mapping_table.csv')
#hgnc.aliases <- hgnc_aliases(hgnc)
#hgnc.aliases <- do.call(rbind, hgnc.aliases)
hgnc.aliases = read.csv( 'data/hgnc_aliases.tsv', sep = '\t')


#-------------------------------------------
# Due to the different report format, we deal
# with Whitehead and broad data seperately

# WHITEHEAD
dfs1 = lapply(paths[paths.whitehead], function(path) {
  
  df = read.csv(path, sep = '\t')
  write(paste0('[WHITEHEAD] loading ', path,'...'),stdout())
  
  df$Accession = as.character(df$Accession)
  df$Description = as.character(df$Description)
  
  ## deal with species
  species = strsplit(df$Accession, '_')
  species = unlist(lapply(species, function(x) x[length(x)]))
  
  # fromm all all data pooled
  ok_species = c('MOUSE','HUMAN','BOVIN','YEAST','RABIT', 'SHEEP', 'ECOLI','HORSE')
  species[!species %in% ok_species] <- NA
  df$species <- species
  
  ## deal with flag-tag by appending a new column with the tag
  tag <- grepl('FLAG',df$Accession)
  df$flag.tag <- NA
  df$flag.tag[tag] <- unlist(lapply(strsplit(df$Accession[tag], split = '\\|'), function(x) x[length(x)]))
  
  ## keratins
  df$grepl.keratin = F
  df$animal.protein = grepl('(Porcine)|(Capra hircus)|(Bos taurus)', df$Description)
  df$grepl.kerantin[grepl('Keratin', df$Description)] <- T
  
  ## deal with gene mapping by looking at the accession and description
  df$gene.description = NA #description.keyword(keyword = 'GN=', from = df$Description)
  df$gene.description[grepl('GN\\=', df$Description)] = description.keyword(keyword = 'GN=', from = df$Description[grepl('GN\\=', df$Description)])
  
  ## get uniprot ID
  df$uniprot.accession.full = accession.uniprot.get(df$Accession)
  
  # get isoforms 
  x = uniprot.isoform.get(df$uniprot.accession.full)
  df$uniprot.accession.id = x$uniprot
  #df$uniprot.accession.isoform = x$id
  
  # use table to map to gene id
  df$gene.aliases.mapped = map_uniprot_to_gene(as.character(df$uniprot.accession.id), uniprot_to_gene)
  df$gene.first = unlist(lapply(strsplit(df$gene.aliases.mapped, split = '\\ '), function(x) x[1]))
  
  # the genes that can't me mapped using UNIPROT
  # will be mapped using the description (given a name)
  df$gene.first.description = df$gene.first
  df$gene.first.description[is.na(df$gene.first.description)] <- df$gene.description[is.na(df$gene.first.description)]
  
  # check whether symbol is found in HGNC approved symbols
  df$gene.final.approved.symbol = df$gene.first.description
  df$gene.final.approved.symbol[!is.na(df$gene.first.description)] = as.logical((df$gene.first.description %in% hgnc$Approved.symbol)[!is.na(df$gene.first.description)])
  df$gene.final = df$gene.first.description
  
  # remove flag tag 
  flag.bool = !is.na(df$flag.tag) & is.na(df$gene.final)
  if (any(flag.bool)){
    df$gene.final[flag.bool] <-  get_flag_bait_gene(df$Accession[flag.bool])
  }
  
  # standardize gene.names
  df$gene.final <- standardize_bait_names(df$gene.final)
  
  # delinate tags
  df$mutant = grepl('\\(Mutant\\)', df$Description)
  df$wildtype = grepl('\\(WT\\)', df$Description)
  df$Uncharacterized <- grepl('Uncharacterized', df$Description)
  
  #newname
  #newpath = gsub('\\.tsv','\\.mapped\\.tsv', path)
  newpath = gsub(writefrom, writeto, path)
  
  stopifnot('gene.final' %in% colnames(df))
  
  write.table(df, file = newpath, sep = '\t', quote = F, row.names = F)
  return(df)
  
})


# BROAD
dfs2 = lapply(paths[paths.broad], function(path) {
  
  # BROAD
  df = read.csv(path, sep = '\t', stringsAsFactors = F)
  write(paste0('[BROAD] loading ', path,'...'),stdout())
  
  # map all all genes and take the first
  if (any(grepl('accession_numbers', colnames(df)))){
    df$genes = unlist(lapply(as.character(df$accession_numbers), function(x) whitehead.accession.uniprot.map(x, uniprot_to_gene)))
    df$gene.first = unlist(lapply(as.character(df$accession_number), function(x) whitehead.accession.uniprot.map(x, uniprot_to_gene)))
  } else {
    df$gene.first = as.character(df$gene)
  }
  if (!'species' %in% colnames(df)){
    df$species <- NA
  }
  
  # Use ID to map rest of genes
  df$gene.final = df$gene.first
  bool = is.na(df$gene.final)
  if (any(bool)){
    df$gene.final[bool] <- toupper(df$id[bool])
  }
  
  # deal with chromosome positions names
  #bool = grepl('^C([0-9]+|(X|Y))orf[0-9]+$', df$gene.final)
  #if (any(bool)){
  #  df$gene.final[bool] <- gsub('ORF','orf',df$gene.final[bool])
  #}
  
  # check whether the final gene is in the HGNC approved symbols
  df$gene.final <- standardize_bait_names(df$gene.final)
  df$gene.final.approved.symbol = df$gene.final
  df$gene.final.approved.symbol[!is.na(df$gene.final)] = as.logical((df$gene.final %in% hgnc$Approved.symbol)[!is.na(df$gene.final)])
  
  #newname
  #newpath = gsub('\\.tsv','\\.mapped\\.tsv', path)
  newpath = gsub(writefrom, writeto, path)
  
  stopifnot('gene.final' %in% colnames(df))
  
  write.table(df, file = newpath, sep = '\t', quote = F, row.names = F)
  
})



# download HGNC symbols

# chromosome build 37
#library(biomaRt)
#ensembl.grch37 <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", 
#                       host    = "grch37.ensembl.org", 
#                       path    = "/biomart/martservice", 
#                       dataset = "hsapiens_gene_ensembl")
#
#filters = listFilters(ensembl.grch37)
#filters[grepl('transcript', filters$name),]
#
#
#
# ref = getBM(attributes=c('hgnc_symbol', 'end','chromosome_name'), 
#      mart = ensembl.grch37)





